src/vaadin-radio-group-mixin.ts

```diff
@@ -0,0 +1,345 @@
+     */
+    @property({ type: Boolean, reflect: true }) invalid: boolean | null | undefined = false;
+
+    /**
+     * Radio group is aligned vertically by default. Set to "horizontal" to change this.
+     */
+    @property({ type: String, reflect: true }) orientation: string | null | undefined;
+
+    /**
+     * When true, the radio group must have a value set by the user.
+     */
+    @property({ type: Boolean, reflect: true }) required: boolean | null | undefined;
+
+    /**
+     * When true, the user cannot modify the value of the radio group.
+     */
+    @property({ type: Boolean, reflect: true }) readonly: boolean | null | undefined;
+
+    /**
+     * Error to show when the radio group value is invalid.
+     */
+    @property({ type: String, attribute: 'error-message' }) errorMessage: string | null | undefined;
+
+    private _boundCheckedChanged = this._onCheckedChanged.bind(this) as EventListener;
+
+    private _checkedButton?: RadioButton;
+
+    private _errorId = `vaadin-radio-group-error-${(uniqueId += 1)}`;
+
+    private _inputChange?: boolean;
+
+    protected render() {
+      return html`
+        <div class="vaadin-group-field-container">
+          <label part="label">${this.label}</label>
+
+          <div part="group-field">
+            <slot id="slot"></slot>
+          </div>
+
+          <div
+            id="${this._errorId}"
+            part="error-message"
+            aria-live="assertive"
+            aria-hidden="${this.errorMessage && this.invalid ? 'false' : 'true'}"
+            ?hidden="${!this.invalid}"
+          >
+            ${this.errorMessage}
+          </div>
+        </div>
+      `;
+    }
+
+    constructor() {
+      super();
+
+      this.addEventListener('change', (event: Event) => {
+        if (this.items.includes(event.composedPath()[0] as HTMLElement)) {
+          event.stopImmediatePropagation();
+          this._inputChange = true;
+        }
+      });
+    }
+
+    /**
+     * Returns true if `value` is valid.
+     */
+    validate(): boolean {
+      this.invalid = !this.checkValidity();
+      return this.invalid === false;
+    }
+
+    /**
+     * Returns true if the current value satisfies all constraints (if any).
+     */
+    checkValidity(): boolean {
+      return !this.required || !!this.value;
+    }
+
+    protected firstUpdated(props: PropertyValues) {
+      super.firstUpdated(props);
+
+      this.setAttribute('role', 'radiogroup');
+    }
+
+    protected updated(props: PropertyValues) {
+      super.updated(props);
+
+      if (props.has('disabled') || props.has('readonly')) {
+        this._updateDisableButtons();
+      }
+
+      if (props.has('label')) {
+        this.toggleAttribute('has-label', Boolean(this.label));
+      }
+
+      if (props.has('value')) {
+        this._valueChanged(this.value, props.get('value'));
+
+        this.dispatchEvent(
+          new CustomEvent('value-changed', {
+            detail: { value: this.value }
+          })
```

Comment:
There should be a way in TS to provide a type for the CustomEvent details object in this kind of case too so we would get type safety for the `detail.value` property here.

Reply:
Maybe we could have predefined generic types/interfaces for detail object which only has a value property of specific type e.g. `DetailValueBoolean`, `DetailValueString`.

Reply:
It seems to work like this:
```suggestion
          new CustomEvent('value-changed', {
            detail: { value: this.value }
          }) as CustomEvent<DetailValueString>
```
including:
```typescript
interface DetailValueString {
  value: string;
}
```

Or even cleaner like this imo:
```suggestion
          new CustomEvent('value-changed', {
            detail: { value: this.value }
          }) as ValueChangedEvent
```

including:
```typescript
interface DetailValueString {
  value: string;
}

type ValueChangedEvent = CustomEvent<DetailValueString>;
```

Reply:
Thanks, actually this needs to be done for all the next generation components.
As mentioned above, here is the tracking issue: https://github.com/vaadin/component-mixins/issues/59

Reply:
Added TODOs with a reference to the issue linked above.